File: dart/test/qr_test.dart
//
//  rerout
//  qr_test.dart
//
//  Official Dart SDK for the Rerout branded-link API.
//
//  Created by Ngonidzashe Mangudya on 20/05/2026.
//  Copyright (c) 2026 Codecraft Solutions
//  Licensed under the MIT License
//  https://codecraftsolutions.co.za
//

import 'package:rerout/rerout.dart';
import 'package:test/test.dart';

void main() {
  group('Rerout.qr.url', () {
    test('returns the basic path with no options', () {
      final r = Rerout.initialize(apiKey: 'rrk_test');
      expect(r.qr.url('q4'), 'https://api.rerout.co/v1/links/q4/qr');
    });

    test('appends every option that is set', () {
      final r = Rerout.initialize(apiKey: 'rrk_test');
      final url = Uri.parse(
        r.qr.url(
          'q4',
          options: const QrOptions(
            size: 12,
            margin: 2,
            ecc: 'H',
            domain: 'go.brand.com',
            refresh: true,
          ),
        ),
      );
      expect(url.path, '/v1/links/q4/qr');
      expect(url.queryParameters['size'], '12');
      expect(url.queryParameters['margin'], '2');
      expect(url.queryParameters['ecc'], 'H');
      expect(url.queryParameters['domain'], 'go.brand.com');
      expect(url.queryParameters['refresh'], '1');
    });

    test('omits unset options', () {
      final r = Rerout.initialize(apiKey: 'rrk_test');
      expect(
        r.qr.url('q4', options: const QrOptions(size: 8)),
        'https://api.rerout.co/v1/links/q4/qr?size=8',
      );
    });

    test('encodes special code characters', () {
      final r = Rerout.initialize(apiKey: 'rrk_test');
      expect(
        r.qr.url('hello world'),
        'https://api.rerout.co/v1/links/hello%20world/qr',
      );
    });

    test('honours a custom baseUrl', () {
      final r = Rerout.initialize(
        apiKey: 'rrk_test',
        baseUrl: 'https://api.staging.example.com',
      );
      expect(
        r.qr.url('q4'),
        'https://api.staging.example.com/v1/links/q4/qr',
      );
    });

    test('accepts a string refresh override', () {
      final r = Rerout.initialize(apiKey: 'rrk_test');
      final url = Uri.parse(
        r.qr.url('q4', options: const QrOptions(refresh: 'v2')),
      );
      expect(url.queryParameters['refresh'], 'v2');
    });
  });
}

File: typescript/test/qr.test.ts
import { describe, expect, it, vi } from 'vitest'
import { Rerout, buildQrUrl } from '../src/index.js'

describe('buildQrUrl', () => {
  it('returns the basic path with no options', () => {
    expect(buildQrUrl({ baseUrl: 'https://api.rerout.co', code: 'q4' })).toBe(
      'https://api.rerout.co/v1/links/q4/qr',
    )
  })

  it('appends every option that is set', () => {
    const url = buildQrUrl({
      baseUrl: 'https://api.rerout.co',
      code: 'q4',
      options: { size: 12, margin: 2, ecc: 'H', domain: 'go.brand.com', refresh: true },
    })
    const parsed = new URL(url)
    expect(parsed.pathname).toBe('/v1/links/q4/qr')
    expect(parsed.searchParams.get('size')).toBe('12')
    expect(parsed.searchParams.get('margin')).toBe('2')
    expect(parsed.searchParams.get('ecc')).toBe('H')
    expect(parsed.searchParams.get('domain')).toBe('go.brand.com')
    expect(parsed.searchParams.get('refresh')).toBe('1')
  })

  it('omits undefined option fields', () => {
    const url = buildQrUrl({
      baseUrl: 'https://api.rerout.co',
      code: 'q4',
      options: { size: 8 },
    })
    expect(url).toBe('https://api.rerout.co/v1/links/q4/qr?size=8')
  })

  it('encodes special code characters', () => {
    const url = buildQrUrl({ baseUrl: 'https://api.rerout.co', code: 'hello world' })
    expect(url).toBe('https://api.rerout.co/v1/links/hello%20world/qr')
  })

  it('strips trailing slashes from baseUrl', () => {
    const url = buildQrUrl({
      baseUrl: 'https://api.rerout.co///',
      code: 'q4',
    })
    expect(url).toBe('https://api.rerout.co/v1/links/q4/qr')
  })

  it('accepts a string refresh override', () => {
    const url = buildQrUrl({
      baseUrl: 'https://api.rerout.co',
      code: 'q4',
      options: { refresh: 'v2' },
    })
    expect(new URL(url).searchParams.get('refresh')).toBe('v2')
  })
})

describe('Rerout.qr', () => {
  it('url() returns the same value as buildQrUrl', () => {
    const r = new Rerout({ apiKey: 'rrk_test', baseUrl: 'https://api.x.dev' })
    expect(r.qr.url('q4', { size: 10 })).toBe(
      buildQrUrl({ baseUrl: 'https://api.x.dev', code: 'q4', options: { size: 10 } }),
    )
  })

  it('svg() fetches the QR endpoint with bearer auth', async () => {
    const fetchImpl = vi.fn(async (input: RequestInfo | URL, init?: RequestInit) => {
      const url = typeof input === 'string' ? input : input.toString()
      expect(url).toBe('https://api.rerout.co/v1/links/q4/qr?size=12')
      const headers = (init?.headers ?? {}) as Record<string, string>
      expect(headers.authorization).toBe('Bearer rrk_test')
      return new Response(JSON.stringify('<svg/>'), {
        status: 200,
        headers: { 'content-type': 'application/json' },
      })
    }) as typeof fetch
    const r = new Rerout({ apiKey: 'rrk_test', fetch: fetchImpl })
    await r.qr.svg('q4', { size: 12 })
  })
})
